### PR TITLE
Rotg

### PR DESCRIPTION
**Pull request by fabiomestre**
Adds support for the BLAS level 1 rotg api (sycnhronous and asynchronous versions).
Adds a unit test and benchmark for rotg

### CODE REVIEW

File: include/blas_meta.h [hunk 1]
```diff
@@ -159,6 +160,13 @@ first_vector_t concatenate_vectors(first_vector_t first_vector,
   return std::move(first_vector);
 }
 
+/**
+ * @brief Defines if a type is a scalar within the context of sycl.
+ * @tparam type The type to be tested.
+ */
+template <typename type>
```
Comment (fabiomestre) [inline include/blas_meta.h:167]: This is still using a C++17 feature (template variables). It compiles but give a warning. 
I seem to be unable to get it working with C++14 (other than copying the condition to all the function declarations)
Open to suggestions :)

File: test/unittest/blas1/blas1_rot_test.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename blas1_rot_test.cpp
+ *
+ **************************************************************************/
+
+#include "blas_test.hpp"
```
Comment (fabiomestre) [inline test/unittest/blas1/blas1_rot_test.cpp:26]: Probably not clear from the PR but this file is the existing unit test for rot operator. It was called rotg even though it didn't test rotg. So I renamed it to blas1_rot_test.cpp

File: src/interface/blas1/rotg_return.cpp.in [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * \brief Synchronous version of rotg.
+ * Given the Cartesian coordinates (a, b) of a point, the rotg routines
+ * return the parameters c, s, r, and z associated with the Givens rotation.
+ * @tparam executor_t Executor type
+ * @tparam scalar_t Scalar type
+ * @param ex Executor
+ * @param a[in, out] On entry, x-coordinate of the point. On exit, the scalar z.
+ * @param b[in, out] On entry, y-coordinate of the point. On exit, the scalar r.
+ * @param c[out] scalar representing the output c.
+ * @param s[out] scalar representing the output s.
```
Comment (hjabird) [inline src/interface/blas1/rotg_return.cpp.in:47]:
```suggestion
 * @param c[out] Scalar representing the output c.
 * @param s[out] Scalar representing the output s.
```

File: src/interface/blas1_interface.hpp [hunk 4]
```diff
@@ -346,6 +346,78 @@ typename executor_t::policy_t::event_t _rot(
...
+ * \brief Synchronous version of rotg.
+ * Given the Cartesian coordinates (a, b) of a point, the rotg routines
+ * return the parameters c, s, r, and z associated with the Givens rotation.
+ * @tparam executor_t Executor type
+ * @tparam scalar_t Scalar type
+ * @param ex Executor
+ * @param a[in, out] On entry, x-coordinate of the point. On exit, the scalar z.
+ * @param b[in, out] On entry, y-coordinate of the point. On exit, the scalar r.
+ * @param c[out] scalar representing the output c.
+ * @param s[out] scalar representing the output s.
```
Comment (hjabird) [inline src/interface/blas1_interface.hpp:393]:
```suggestion
 * @param c[out] Scalar representing the output c.
 * @param s[out] Scalar representing the output s.
```

File: cmake/CmakeFunctionHelper.cmake [hunk 5]
```diff
@@ -367,6 +367,91 @@ add_sycl_to_target(TARGET ${func} SOURCES ${FUNC_SRC})
 endfunction(generate_blas_ternary_objects)
 
 
+# blas function for generating source code for the rotg operator (asynchronous version with containers)
+function(generate_blas_rotg_objects blas_level func)
+  set(LOCATION "${SYCLBLAS_GENERATED_SRC}/${blas_level}/${func}/")
+  foreach (executor ${executor_list})
+    foreach (data ${data_list})
+      cpp_type(cpp_data ${data})
+      set(container_list_in)
```
Comment (hjabird) [inline cmake/CmakeFunctionHelper.cmake:376]: This doesn't seem to get used.

File: src/operations/blas1_trees.hpp [hunk 6]
```diff
@@ -538,6 +538,87 @@ AssignReduction<operator_t, lhs_t, rhs_t>::adjust_access_displacement() {
...
+      abs_a > abs_b ? SignOperator::eval(a_ref): SignOperator::eval(b_ref);
+  const value_t r =
+      ProductOperator::eval(sigma, HypotenuseOperator::eval(a_ref, b_ref));
+
+
+  if (r == zero::value()) {
+    c_ref = one::value();
+    s_ref = zero::value();
+  } else {
+    c_ref = DivisionOperator::eval(a_ref, r);
```
Comment (hjabird) [inline src/operations/blas1_trees.hpp:574]: Do we need this if abs_1 >= abs_b? Can we defer calculating it?

**Review by hjabird — COMMENTED** [review]
Lots of code, but very little to comment on. Nice work.

File: include/blas_meta.h [hunk 7]
```diff
@@ -159,6 +160,13 @@ first_vector_t concatenate_vectors(first_vector_t first_vector,
...
+constexpr bool is_sycl_scalar = std::is_scalar<type>::value || std::is_same<cl::sycl::half, type>::value;
```
[the remark this answers is not in the document]
Comment (fabiomestre) [reply include/blas_meta.h]: Thanks for the suggestion @hjabird! That works well :)

File: src/operations/blas1_trees.hpp (hunk 6 above)
Comment (fabiomestre) [reply src/operations/blas1_trees.hpp:574]: Not sure if I understand the suggestion. If abs_a >= abs_b, the only thing that changes is the z output (parameter b) and the sigma.  But we still need to calculate a, c and s output parameters.

Comment (hjabird) [reply src/operations/blas1_trees.hpp:574]: I missed that `c` and `s` where also returned.

File: include/blas_meta.h (hunk 1 above)
Comment (Rbiessy) [reply include/blas_meta.h:167]: What is the warning exactly? `std::is_scalar` and `std::conditional` both seem to be present in C++11.

File: test/unittest/blas1/blas1_rot_test.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  fill_random(a_v);
+  std::vector<data_t> b_v(size * incY);
+  fill_random(b_v);
+
+  // Output vectors
+  std::vector<data_t> out_s(1, 10.0);
+  std::vector<data_t> a_cpu_v(a_v);
+  std::vector<data_t> b_cpu_v(b_v);
+
+  // Looks like we don't have a SYCL rotg implementation
```
Comment (Rbiessy) [inline test/unittest/blas1/blas1_rot_test.cpp:52]: That comment isn't valid anymore.

File: test/unittest/blas1/blas1_rotg_test.cpp [hunk 9]
```diff
@@ -26,84 +26,81 @@
...
     const ::testing::TestParamInfo<combination_t<T>>& info) {
-  int size, incX, incY;
-  BLAS_GENERATE_NAME(info.param, size, incX, incY);
+  api_type api;
+  T a, b;
+  BLAS_GENERATE_NAME(info.param, api, a, b);
 }
 
-BLAS_REGISTER_TEST_ALL(Rotg, combination_t, combi, generate_name);
+BLAS_REGISTER_TEST_ALL(Rotg, combination_t, combi, generate_name);
```
Comment (Rbiessy) [inline test/unittest/blas1/blas1_rotg_test.cpp:106]: Missing new line at the end of the file.

**Review by Rbiessy — COMMENTED** [review]
I made some minor comments but it looks good to me!

File: test/unittest/blas1/blas1_rot_test.cpp (hunk 8 above)
Comment (fabiomestre) [reply test/unittest/blas1/blas1_rot_test.cpp]: Fixed

File: test/unittest/blas1/blas1_rotg_test.cpp (hunk 9 above)
Comment (fabiomestre) [reply test/unittest/blas1/blas1_rotg_test.cpp]: Fixed

File: include/blas_meta.h (hunk 1 above)
Comment (fabiomestre) [reply include/blas_meta.h:167]: The code already has the changes suggested by @hjabird . Before I was using variable templates which are C++17 only feature

**Review by Rbiessy — APPROVED** [review]

**Review by hjabird — APPROVED** [review]
Changes look good!

**Review by KumudhaN — APPROVED** [review]

**Pull request merged**